### PR TITLE
fix: correctly load timestamps from tables in artifacts

### PR DESCRIPTION
**Pull request by dannygoldstein**
Fixes [WB-9763](https://wandb.atlassian.net/browse/WB-9763)

Description
-----------
Tables with timestamps logged as cells of a column did not deserialize properly because on our backend, we store timestamps as floats and not instances of datetime. This caused our type system in client to think that they were numbers. So when we tried to deserialize them as timestamps we would see:

```
In [6]: train_table = artifact.get("train")
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [6], in <cell line: 1>()
----> 1 train_table = artifact.get("train")

File ~/.pyenv/versions/3.9.7/envs/weave_internal/lib/python3.9/site-packages/wandb/apis/public.py:3841, in Artifact.get(self, name)
   3839 with open(item_path) as file:
   3840     json_obj = json.load(file)
-> 3841 result = wb_class.from_json(json_obj, self)
   3842 result._set_artifact_source(self, name)
   3843 return result

File ~/.pyenv/versions/3.9.7/envs/weave_internal/lib/python3.9/site-packages/wandb/data_types.py:589, in Table.from_json(cls, json_obj, source_artifact)
    584 if column_types is not None:
    585     dtypes = [
    586         column_types.params["type_map"][col] for col in json_obj["columns"]
    587     ]
--> 589 new_obj = cls(columns=json_obj["columns"], data=data, dtype=dtypes)
    591 if column_types is not None:
    592     new_obj._column_types = column_types

File ~/.pyenv/versions/3.9.7/envs/weave_internal/lib/python3.9/site-packages/wandb/data_types.py:279, in Table.__init__(self, columns, data, rows, dataframe, dtype, optional, allow_mixed_types)
    277         self._init_from_dataframe(data, columns, optional, dtype)
    278     else:
--> 279         self._init_from_list(data, columns, optional, dtype)
    281 # legacy
    282 elif rows is not None:

File ~/.pyenv/versions/3.9.7/envs/weave_internal/lib/python3.9/site-packages/wandb/data_types.py:306, in Table._init_from_list(self, data, columns, optional, dtype)
    304 self._make_column_types(dtype, optional)
    305 for row in data:
--> 306     self.add_data(*row)

File ~/.pyenv/versions/3.9.7/envs/weave_internal/lib/python3.9/site-packages/wandb/data_types.py:470, in Table.add_data(self, *data)
    463         self.cast(
    464             self.columns[ndx],
    465             _dtypes.TypeRegistry.type_of(item),
    466             optional=False,
    467         )
    469 # Update the table's column types
--> 470 result_type = self._get_updated_result_type(data)
    471 self._column_types = result_type
    473 # rows need to be mutable

File ~/.pyenv/versions/3.9.7/envs/weave_internal/lib/python3.9/site-packages/wandb/data_types.py:491, in Table._get_updated_result_type(self, row)
    489 result_type = current_type.assign(incoming_row_dict)
    490 if isinstance(result_type, _dtypes.InvalidType):
--> 491     raise TypeError(
    492         "Data row contained incompatible types:\n{}".format(
    493             current_type.explain(incoming_row_dict)
    494         )
    495     )
    496 return result_type

TypeError: Data row contained incompatible types:
{'p (mbar)': 0.9446036736514699, 'T (degC)': -1.9829810037354159, 'Tpot (K)': -2.0423569129810173, 'Tdew (degC)': -1.918783022245483, 'rh (%)': 1.117756372932424, 'VPmax (mbar)': -1.3030968262245533, 'VPact (mbar)': -1.4766629860307678, 'VPdef (mbar)': -0.7906578632181194, 'sh (g/kg)': -1.4793450725529895, 'H2OC (mmol/mol)': -1.482007271939327, 'rho (g/m**3)': 2.218807581641912, 'Wx': 0.19578502792871674, 'Wy': 0.22130537565656638, 'max Wx': 0.11282737985058558, 'max Wy': 0.21759103699023694, 'Day sin': 0.36608513953959493, 'Day cos': 1.3660141911643051, 'Year sin': -0.06104243564351187, 'Year cos': 1.428406307008361, 'Date Time': 1230800400000.0, 'timestamp': 1230771600.0} of type {'p (mbar)': Number, 'T (degC)': Number, 'Tpot (K)': Number, 'Tdew (degC)': Number, 'rh (%)': Number, 'VPmax (mbar)': Number, 'VPact (mbar)': Number, 'VPdef (mbar)': Number, 'sh (g/kg)': Number, 'H2OC (mmol/mol)': Number, 'rho (g/m**3)': Number, 'Wx': Number, 'Wy': Number, 'max Wx': Number, 'max Wy': Number, 'Day sin': Number, 'Day cos': Number, 'Year sin': Number, 'Year cos': Number, 'Date Time': Number, 'timestamp': Number} is not assignable to {'p (mbar)': None or None or Number, 'T (degC)': None or None or Number, 'Tpot (K)': None or None or Number, 'Tdew (degC)': None or None or Number, 'rh (%)': None or None or Number, 'VPmax (mbar)': None or None or Number, 'VPact (mbar)': None or None or Number, 'VPdef (mbar)': None or None or Number, 'sh (g/kg)': None or None or Number, 'H2OC (mmol/mol)': None or None or Number, 'rho (g/m**3)': None or None or Number, 'Wx': None or None or Number, 'Wy': None or None or Number, 'max Wx': None or None or Number, 'max Wy': None or None or Number, 'Day sin': None or None or Number, 'Day cos': None or None or Number, 'Year sin': None or None or Number, 'Year cos': None or None or Number, 'Date Time': None or None or Timestamp, 'timestamp': None or None or Number}
Key 'Date Time':
	Number not assignable to None or None or Timestamp
		Number not assignable to None
	and
		Number not assignable to None
	and
		Number not assignable to Timestamp
```

This would cause `use_artifact().get()` to fail anytime we wanted to `.get()` a table with datetime cells.

This fixes the problem by adding a special deserialization path for timestamps in `wandb.Table.from_json`

Testing
-------
Unit test.

Checklist
-------
 - [x] Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: wandb/data_types.py [hunk 1]
```diff
@@ -549,6 +550,14 @@ def from_json(cls, json_obj, source_artifact):
                     for t in col_type.params["allowed_types"]:
                         if isinstance(t, _dtypes.NDArrayType):
                             ndarray_type = t
+                        elif isinstance(t, _dtypes.TimestampType):
+                            timestamp_column_indices.append(
+                                json_obj["columns"].index(col_name)
+                            )
+
+                elif isinstance(col_type, _dtypes.TimestampType):
+                    timestamp_column_indices.append(json_obj["columns"].index(col_name))
```
Comment (dmitryduev) [inline wandb/data_types.py:559]:
```suggestion
                    timestamp_column_indices.add(json_obj["columns"].index(col_name))
```

Comment (dmitryduev) [reply wandb/data_types.py:559]: Please add a test for this case.

File: wandb/data_types.py [hunk 2]
```diff
@@ -569,7 +578,14 @@ def from_json(cls, json_obj, source_artifact):
         for r_ndx, row in enumerate(json_obj["data"]):
             row_data = []
             for c_ndx, item in enumerate(row):
-                cell = item
+                cell = (
+                    item
+                    if c_ndx not in timestamp_column_indices or not item
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply wandb/data_types.py:583]: actually, why did you need `not item`?

[the remark this answers is not in the document]
Comment (dannygoldstein) [reply wandb/data_types.py]: because we dont want to try to create a timestamp from `None`

[the remark this answers is not in the document]
Comment (dannygoldstein) [reply wandb/data_types.py]: shoot, this should actually be `IS none`

[the remark this answers is not in the document]
Comment (dannygoldstein) [reply wandb/data_types.py]: sorry, it was right before. the suggestion flipped the bit ! will update...

File: wandb/data_types.py [hunk 3]
```diff
@@ -536,6 +536,7 @@ def from_json(cls, json_obj, source_artifact):
         data = []
         column_types = None
         np_deserialized_columns = {}
+        timestamp_column_indices = []
```
[the remark this answers is not in the document]
Comment (dannygoldstein) [reply wandb/data_types.py]: i think this actually creates an empty dict?

**Comment by dannygoldstein** [comment]
had a nice chat with @tssweeney about this PR. we discussed the fact that currently both numbers and date times serialize to numbers so there can be a degeneracy when a table column typed as union[datetime, number] in an artifact has an has an ambiguous deserialized type

we decided that an approach to fixing this would be to add a new WBType for timestamp and update the corresponding type definitions in cg. but we agreed that the approach here is safer for now as it has a smaller surface area: it needs no changes in cg or the ui. 

there is still an edge case here for the union[datetime, number] case, but this is an edge case which we are content to have be a "known bug" for now and wait until there is a broader investment in dataset management to fix. 

we should do a datatype seraizliation format audit to identify these types of degeneracies and improve them in weave python.

**Review by tssweeney — APPROVED** [review]

**Pull request merged**